internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+	}
+	return nil
+}
+
+// createIDCert create a public x509 CA certificate, signed by the identity
+// key. This certificate is included in the TLS certificate chain as the
+// non-leaf certificate, allowing the client to pin this certificate during
+// the client server trust exchange (pairing) procedure.
+func createIDCert(signer crypto.Signer) (*x509.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	template := &x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             systemTime(),
+		NotAfter:              systemTime().Add(idCertValidity),
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		// We can only sign leaf certificates with this.
+		MaxPathLen:     0,
+		MaxPathLenZero: true,
+	}
+
+	// The identity CA cert is self-signed, which allows a client to verify the
+	// incoming version of this with the pinned version.
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, signer.Public(), signer)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// isCertActive checks of the current time now is within the start and end time
+// validity period as defined by the certificate.
+func isCertActive(cert *x509.Certificate) bool {
+	now := systemTime()
+	// Note that we shorten the NotAfter timestamp by a tlsCertRenewWindow
+	// duration to avoid a TLS handshake race against the expiry time.
+	if now.Before(cert.NotBefore) || now.After(cert.NotAfter.Add(-tlsCertRenewWindow)) {
+		return false
+	}
+	return true
+}
+
+// isCertDerived checks if the public key in the certificate matches
+// the signer public key. This confirms the certificate matches the
+// identity key.
+func isCertDerived(cert *x509.Certificate, signer crypto.Signer) bool {
+	signerPublic := signer.Public().(ed25519.PublicKey)
+	certPublic := cert.PublicKey.(ed25519.PublicKey)
+	return bytes.Equal(signerPublic, certPublic)
+}
+
+// expectPermission return an error if the specified directory or file
+// path is not matching the supplied permissions.
+func expectPermission(path string, perm fs.FileMode) error {
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	actualPerm := info.Mode().Perm()
+	if actualPerm != perm {
+		return fmt.Errorf("expected permission 0o%o (got 0o%o) for %q", perm, actualPerm, path)
+	}
+	return nil
+}
+
+// pathExists returns true of the path exists, false if it does not. If the
+// operation reports an unrelated error, the error is returned.
+func pathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
```

Comment:
Per the docs, this predates `errors.Is`, and new code should use `errors.Is(err, fs.ErrNotExist)` (similar in the other PR).

Reply:
Fixed. Resolving.